add prometheus library

mustafaabdullahk opened the pull request:
# Hey Zephyr Community! 🚀 

I'm excited to announce the launch of our latest addition to the Zephyr ecosystem: the Zephyr Prometheus Library! 🎉

## What is Zephyr Prometheus Library?
Zephyr Prometheus is a lightweight library designed to seamlessly integrate Prometheus metrics into your Zephyr projects. Whether you're developing for IoT, edge computing, or any other embedded application with Zephyr, monitoring and collecting metrics is now easier than ever.

## Features:
📊 Effortless Integration: With just a few lines of code, you can start exposing custom metrics from your Zephyr applications.

🔍 Flexible Metrics Collection: Monitor crucial aspects of your application's performance, such as memory usage, task execution times, and custom-defined metrics.

🔒 Secure Communication: Ensure the integrity and confidentiality of your metrics data with secure communication protocols supported by Zephyr.

we'r talk about that pr following discussion: #72127

## Metric Server

![prom-metrics-screen](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/03b7baa9-e78e-476a-b735-c99861ad2150)

## Grafana

![grafana-screen](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/2ff481f8-d963-40bc-bbda-4b7267ebf977)


Signed-off-by: Mustafa Abdullah Kus [mustafa.kus@sparsetechnology.com](mailto:mustafa.kus@sparsetechnology.com)

bjarki-andreasen reviewed (commented):
Looks good, some notes:
* The `pm_` namespace is used for "power management", maybe update it to `net_pm`, or `pmts_`? or simply `prometheus_`
* I think prometheous may be better placed in `subsys/mgmt` alongside integrations like mcumgr, rather than `net`, where we have MQTT, IPV4 etc

mustafaabdullahk commented:
> Looks good, some notes:
> 
>     * The `pm_` namespace is used for "power management", maybe update it to `net_pm`, or `pmts_`? or simply `prometheus_`
> 
>     * I think prometheous may be better placed in `subsys/mgmt` alongside integrations like mcumgr, rather than `net`, where we have MQTT, IPV4 etc

* i will change with prometheus instead of others, first of all i did think that name but give up that name because its too long. if that namespace conflict with power management, should be change. i guess best way change with prometheus namespace.
* i dont sure about place. mr @jukkar (https://github.com/zephyrproject-rtos/zephyr/discussions/72127) did suggest to me that place. if best place, i can change with gladly :)

jukkar commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,69 @@
+/*
```
Minor nit, but the commit message does not follow English grammar and looks bad. Could you fix that by starting sentences with capital letter etc.

jukkar commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,69 @@
+
+#include <stddef.h>
+
+typedef struct {
```
We normally use typedefs only for function pointers in Zephyr and avoid using typedefs for structs. So with this in mind you could just have `struct pm_collector_node { ... }` here. Similar comment to all the other typedefs you have in the header files.

Also, all the structs and their fields need to be documented if they are considered to be "public" i.e., used by the application. For "private" structs, you should use `/** @cond INTERNAL_HIDDEN */` ... `/** @endcond */` to prevent the struct from showing in the generated documentation. See examples in other header files.

jukkar commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,69 @@
+ *
+ * @return Pointer to the created collector.
+ */
+pm_collector_t *pm_collector_create(void);
```
About the `pm_` prefix, I think we could use `prometheus_` as suggested by @bjarki-trackunit. It is a bit longish but we avoid confusion with power mgmt APIs.

jukkar commented on subsys/net/lib/Kconfig:
```diff
@@ -47,4 +47,6 @@ source "subsys/net/lib/trickle/Kconfig"
 
```
The commit `net: lib: kconfig: add prometheus library kconfig source` could be squashed to the commit that introduce the library sources, no need for a separate commit for that.

jukkar commented on subsys/net/lib/CMakeLists.txt:
```diff
@@ -16,6 +16,7 @@ add_subdirectory_ifdef(CONFIG_NET_ZPERF              zperf)
 add_subdirectory_ifdef(CONFIG_NET_SHELL              shell)
 add_subdirectory_ifdef(CONFIG_NET_TRICKLE            trickle)
 add_subdirectory_ifdef(CONFIG_NET_DHCPV6             dhcpv6)
+add_subdirectory_ifdef(CONFIG_PROMETHEUS             prometheus)
```
Also this should be squashed to the commit introducing the library.

jukkar commented on samples/net/prometheus/Kconfig:
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) 2024 Mustafa Abdullah Kus, Sparse Technology
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "prometheus sample application"
```
Better `Prometheus ...`

jukkar commented:
About the location, I initially thought the `net` for this but as @bjarki-trackunit mentioned, the `mgmt` is also possible. I do not have strong opinion about this, what people in the review list think?

mustafaabdullahk commented:
> About the location, I initially thought the `net` for this but as @bjarki-trackunit mentioned, the `mgmt` is also possible. I do not have strong opinion about this, what people in the review list think?

I waiting to opinion for re-place library components. For now i did solve @jukkar reviewed points.

jukkar commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,71 @@
+
+#include <stddef.h>
+
+struct prometheus_collector_node {
+	struct prometheus_metric *metric;
+	struct pm_collector_node *next;
+};
+
+struct prometheus_collector {
+	struct prometheus_collector_node *head;
+	size_t size;
+};
```
If these are part of the API that applications will use, then these needs to be documented so that the generated documentation contains useful information. Same comment to all structs in the header files.

mustafaabdullahk replied:
I will add detailed comments.

jukkar commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,71 @@
+
+struct prometheus_collector_node {
+	struct prometheus_metric *metric;
+	struct pm_collector_node *next;
```
Instead of our own list pointers, would it make sense to use the slist API here, WDYT?

mustafaabdullahk replied:
I wanted to use slist apis at first. I will update with slist instead of my linked list implementation but I should be try. We should will measure overhead between implementations.

jukkar commented on subsys/net/lib/prometheus/pm_collector.c:
```diff
@@ -0,0 +1,134 @@
+	struct prometheus_collector_node *current = collector->head;
+
+	while (current != NULL) {
+		if (strcmp(current->metric->name, name) == 0) {
```
Please use `strncmp()` to make sure we do not read past the end of buffer

jukkar commented on subsys/net/lib/prometheus/pm_formatter.c:
```diff
@@ -0,0 +1,218 @@
+static int write_metric_to_buffer(char *buffer, size_t buffer_size, const char *format, ...)
+{
+	/* helper function to write formatted metric to buffer */
+	char temp_buffer[256];
```
This is quite large and will use quite large amount of stack, do we need this long tmp buffer?

jukkar reviewed (commented):
Did not finish the review yet, but some preliminary comments added.
As a general note I am not a big fan of the dynamic memory we are using all over the place. We have been trying to minimize the usage of malloc in the network stack, but this is not possible always but we should try to minimize malloc use (because if malloc is not used, then there is no possibility to memory leaks).

jukkar commented:
Also we will need some tests for these, could be placed for example to `tests/net/lib/prometheus`

mustafaabdullahk commented:
> Also we will need some tests for these, could be placed for example to `tests/net/lib/prometheus`

We did decide about library place? Can i integrate to under net/lib for now ?

jukkar commented:
> We did decide about library place? Can i integrate to under net/lib for now ?

I think there is no consensus yet, it should be trivial change if the location is changed, just some `git mv`

jukkar commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,73 @@
+ * This structure defines a Prometheus collector.
+ */
+struct prometheus_collector {
+	/* Name of the collector */
+	const char *name;
+	/* Array of metrics associated with the collector */
+	struct prometheus_metric *metric[CONFIG_PROMETHEUS_MAX_METRICS];
+	/* Number of metrics associated with the collector */
```
Please change the comments here to doxygen comments so that the documentation is generated automatically for these. So `/** .... */`

jukkar commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,73 @@
+ *
+ * @param collector Pointer to the collector to register the metric with.
+ * @param metric Pointer to the metric to register.
+ */
```
Return values need to be documented too, otherwise doxygen will complain.

jukkar commented on include/zephyr/net/prometheus/pm_counter.h:
```diff
@@ -0,0 +1,67 @@
+ * * See https://prometheus.io/docs/concepts/metric_types/#counter
+ */
+struct prometheus_counter {
+	/* Base of the Prometheus counter metric */
+	struct prometheus_metric *base;
+	/* Value of the Prometheus counter metric */
```
Change to doxygen comments

jukkar commented on include/zephyr/net/prometheus/pm_gauge.h:
```diff
@@ -0,0 +1,61 @@
+#include <zephyr/sys/iterable_sections.h>
+#include <zephyr/net/prometheus/pm_metric.h>
+
+struct prometheus_gauge {
```
Documentation header missing for the struct, also doxygen comments missing for the fields.

jukkar commented on include/zephyr/net/prometheus/pm_histogram.h:
```diff
@@ -0,0 +1,87 @@
+
+#include <stddef.h>
+
+struct prometheus_histogram_bucket {
```
same here

jukkar commented on include/zephyr/net/prometheus/pm_histogram.h:
```diff
@@ -0,0 +1,87 @@
+ * * See https://prometheus.io/docs/concepts/metric_types/#histogram
+ */
+struct prometheus_histogram {
+	/* Base of the Prometheus histogram metric */
+	struct prometheus_metric *base;
+	/* Array of buckets in the histogram */
+	struct prometheus_histogram_bucket *buckets;
+	/* Number of buckets in the histogram */
+	size_t num_buckets;
+	/* Sum of all observed values in the histogram */
+	double sum;
+	/* Total count of observations in the histogram */
+	unsigned long count;
```
doxygen comments

jukkar commented on include/zephyr/net/prometheus/pm_label.h:
```diff
@@ -0,0 +1,30 @@
+#ifndef ZEPHYR_INCLUDE_PM_LABEL_H_
+#define ZEPHYR_INCLUDE_PM_LABEL_H_
+
+/* maximum length of label key */
+#define MAX_LABEL_KEY_LENGTH   16
+/* maximum length of label value */
+#define MAX_LABEL_VALUE_LENGTH 16
+/* maximum namber of labels per metric */
+#define MAX_LABELS_PER_METRIC  5
+
+struct prometheus_label {
+	/* Prometheus metric label key */
+	char key[MAX_LABEL_KEY_LENGTH];
+	/* Prometheus metric label value */
```
Please change to `/** .. */`

jukkar commented on include/zephyr/net/prometheus/pm_metric.h:
```diff
@@ -0,0 +1,53 @@
+#include <zephyr/sys/iterable_sections.h>
+#include <zephyr/net/prometheus/pm_label.h>
+
+enum prometheus_metric_type {
```
doxygen comments missing

jukkar commented on include/zephyr/net/prometheus/pm_metric.h:
```diff
@@ -0,0 +1,53 @@
+ * to control the metric access and usage.
+ */
+struct prometheus_metric {
+	/* Type of the Prometheus metric. */
```
here too

jukkar commented on include/zephyr/net/prometheus/pm_summary.h:
```diff
@@ -0,0 +1,81 @@
+
+#include <stddef.h>
+
+struct prometheus_summary_quantile {
```
same here

jukkar commented on include/zephyr/net/prometheus/pm_summary.h:
```diff
@@ -0,0 +1,81 @@
+	double value;
+};
+
+struct prometheus_summary {
```
and here

jukkar reviewed (commented):
Thanks, looks pretty good now. I did spot some issues with the documentation that needs to be fixed.

bjarki-trackunit commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,77 @@
+	/** Name of the collector */
+	const char *name;
+	/** Array of metrics associated with the collector */
+	struct prometheus_metric *metric[CONFIG_PROMETHEUS_MAX_METRICS];
```
See https://github.com/zephyrproject-rtos/zephyr/blob/598ab6b39a5e38466a286fb317f39bcd34b8ad20/subsys/modem/modem_stats.c#L26-L31 and https://github.com/zephyrproject-rtos/zephyr/blob/598ab6b39a5e38466a286fb317f39bcd34b8ad20/include/zephyr/modem/stats.h#L18-L23 :)

mustafaabdullahk replied:
I was saw it. I want to integrate that way but first version maybe fixed size array :). I can update that part when write the library test :)

bjarki-trackunit commented on include/zephyr/net/prometheus/pm_collector.h:
```diff
@@ -0,0 +1,77 @@
+ * @brief Prometheus collector APIs.
+ */
+
+#ifndef ZEPHYR_INCLUDE_PM_COLLECTOR_H_
```
The header files don't need to be namespaced if the namespace is in the folder name:
```
#include <zephyr/net/prometheus/pm_collector.h>
```
can be shortened to
```
#include <zephyr/net/prometheus/collector.h>
```
as `pm` / `prometheus` is already in the path :)

bjarki-trackunit replied:
The include guard should also be updated to use the namespace `PROMETHEUS` instead of `PM`
```
#ifndef ZEPHYR_INCLUDE_PROMETHEUS_COLLECTOR_H_
```

mustafaabdullahk replied:
Now we'r talking :). I did change right now

bjarki-trackunit replied:
Nice!

bjarki-trackunit commented on subsys/net/lib/prometheus/Kconfig:
```diff
@@ -0,0 +1,32 @@
+	bool "Prometheus Client Library (Pull Method)"
+	depends on NET_SOCKETS
+	depends on HTTP_SERVER
+	depends on HEAP_MEM_POOL_SIZE != 0
```
If heap is required, you can create a special Kconfig which specifies how much PROMETHEUS requires, which will be summed up by the build system :)

```
if PROMETHEUS

config HEAP_MEM_POOL_ADD_SIZE_PROMETHEUS
        int "Heap size required by prometheus library in bytes"
        default 2048
        range 256 65535

endif # PROMETHEUS
```

You can then remove `depends on HEAP_MEM_POOL_SIZE != 0` as this becomes explicit

mustafaabdullahk replied:
I did remove it. I did missed because before force-pushed, the library use memory pool now the metrics create compile time :). Maybe i can add dynamic metric create feature with that config

bjarki-trackunit replied:
Perfect :)

rlubos commented on samples/net/prometheus/src/main.c:
```diff
@@ -0,0 +1,240 @@
+		return rc;
+	}
+
+	rc = sendall(client, hex_len, rc);
```
I'm a bit puzzled here, why are we sending the reply back to the client directly from the resource handler? Wouldn't that be in conflict with the server code, which actually is responsible for sending the reply back to the client?

I would expect, that the application callback simply fills the `buffer` with the response payload, and return the number of bytes written back to the server. The server should then send them back in a HTTP "chunk".

Or am I missing something here? For reference:
https://docs.zephyrproject.org/latest/connectivity/networking/api/http_server.html#dynamic-resources

mustafaabdullahk replied:
I'm totally aggree with you. If review the `dynamic_get_req` callback you will see that part:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/3eec4384-86b1-4c3a-b2b7-2dd4241fd00c)

When I try to copy `dynamic_get_req` buffer callback hang infinite loop because callback return check one time.

rlubos replied:
The callback needs to return 0 to break the loop, that tells the server side that no more data is expected to be included in the response. So in a simple case, I imagine callback gets called twice, the first time you fill out the buffer and return number of bytes written, the second just return 0 indicating that no more data is to be expected.

mustafaabdullahk replied:
I did try first is like your tell and noticed to written size not passing to callback so cant check how many bytes are sent.

rlubos replied:
The server guarantees to send all bytes that have been returned by the callback:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/http/http_server_http1.c#L150

mustafaabdullahk replied:
Yes, you'r right. I try to say the server pass `copy_len` argument to callback. The server calculate `copy_len` size zero every time: https://github.com/zephyrproject-rtos/zephyr/blob/62893328293f8719f72b0a4e2edcb48fef3e125c/samples/net/sockets/http_server/src/main.c#L61

Our scenario following steps:

- Call `dynamic_get_req` with parameters `copy_len` is 0.
- We'r copy our buffer to the server buffer and return buffer size.
- The server perform callback again because `send_len` is bigger than 0.
- We give `copy_len`  argument as zero and copy buffer again.

How do we check that situation?

mustafaabdullahk replied:
I think I figured it out :). CI issues and rebase done.

rlubos reviewed (commented):
Looks pretty neat overall, just one unclarity about the HTTP server operation.

I wonder, shall we add a new entry for the library/sample in `MAINTAINERS.yml` file?

mustafaabdullahk commented:
Hi @jukkar 

I did add tests. Should i generate coverage report and expand test coverage? Is it enough?

bjarki-andreasen commented:
Some tests are definitely malformed :) check `collector/prj.conf` for example.

Try building every test locally, I recommend building each test for native_sim and running it locally (from workspace dir):
```
~/workspace west build -p -b native_sim zephyr/tests/net/lib/prometheus/counter/
~/workspace ./build/zephyr/zephyr.exe
```
Then once all tests seem to build and run as expected, run twister locally to build and run the tests for all supported platforms:
```
~/workspace ./zephyr/scripts/twister -T zephyr/tests/net/lib/prometheus/
```

Counting on CI is way to slow (and wasteful)

josuah commented on samples/net/prometheus/src/certificate.h:
```diff
@@ -0,0 +1,50 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __CERTIFICATE_H__
+#define __CERTIFICATE_H__
```
Not sure how much that matters, but it looked like you used the full path in the other #ifdef guards:
https://github.com/zephyrproject-rtos/zephyr/pull/72804/files#diff-3604770612d200b009418873ab10cdfc46a0c5ba37407de112ba896c0b6a4273R13

kartben replied:
Still applies, I think

mustafaabdullahk replied:
I can't remember what will i do. I just copied from :

https://github.com/zephyrproject-rtos/zephyr/blob/5e2203a6cdd0daf395bb3825d416843ecdc40629/samples/net/sockets/http_server/src/certificate.h#L7. Can you explain more?

josuah replied:
Yes, glad to! It looks like you had it done in one way [here](https://github.com/zephyrproject-rtos/zephyr/pull/72804/files#diff-3604770612d200b009418873ab10cdfc46a0c5ba37407de112ba896c0b6a4273R7-R8):
```
include/zephyr/prometheus/label.h   <- name of the file
ZEPHYR_INCLUDE_PROMETHEUS_LABEL_H_  <- name of the #ifdef guard
```

And in another way for this file `certificate.h`, but I forgot it was a sample, and there does not seems to be a very rigid way for samples:

```
~/zephyrproject/zephyr$ find samples/ -name '*.h' -exec grep -h '#ifndef ' {} + | sort -u | head -n 30
#ifndef APPLICATION_BATTERY_H_
#ifndef APP_SHARED_H
#ifndef APP_SYSCALL_H
#ifndef BYTE
#ifndef COMMON_H__
#ifndef CO_OD_H_
#ifndef DISPLAY_7SEG_HEADER
#ifndef ENC_H
#ifndef EXT_LOG_SYSTEM_ADAPTER_H
#ifndef EXT_LOG_SYSTEM_H
#ifndef FEEDBACK_H_
#ifndef MAIN_H
#ifndef PROD_CONSUMER_APP_A_H
#ifndef PROD_CONSUMER_APP_B_H
#ifndef RETAINED_H_
#ifndef SAMPLE_INSTANCE_H
#ifndef SAMPLE_MODULE_H
#ifndef STREAM_LC3_H
#ifndef STREAM_TX_H
#ifndef ST_BLE_SENSOR_BUTTON_SVC_H_
#ifndef ST_BLE_SENSOR_LED_SVC_H_
#ifndef TENSORFLOW_LITE_MICRO_EXAMPLES_HELLO_WORLD_CONSTANTS_H_
#ifndef TENSORFLOW_LITE_MICRO_EXAMPLES_HELLO_WORLD_MAIN_FUNCTIONS_H_
#ifndef THREAD_DEF_H
#ifndef ZEPHYR_FAKE_DRIVER_H
#ifndef ZEPHYR_SAMPLES_DRIVERS_LED_WS2812_F070RB_BINDINGS_H
#ifndef ZEPHYR_SAMPLES_DRIVERS_LED_WS2812_H_
#ifndef ZEPHYR_SAMPLES_SENSOR_SENSOR_SHELL_INCLUDE_TRIGGER_H
#ifndef ZEPHYR_SAMPLES_SUBSYS_USB_COMMON_SAMPLE_USBD_H
#ifndef _APP_GPIO_H
```

Nevermind then!

josuah commented on samples/net/prometheus/src/https-server-cert.der:
How about using `pem` formatted files which are a bit more git friendly?
Maybe there are good reasons to use DER instead of PEM though...

Or maybe even generating them at every build?

jukkar replied:
The existing network samples are using DER in order to avoid including PEM parser which uses more memory.

josuah replied:
This makes a lot more sense to have DER in the firmware of course. Thanks for the clarification!

josuah commented on samples/net/prometheus/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+   CONFIG_MBEDTLS_KEY_EXCHANGE_PSK_ENABLED AND
+   (CONFIG_NET_SAMPLE_PSK_HEADER_FILE STREQUAL "dummy_psk.h"))
+  add_custom_target(development_psk
+    COMMAND ${CMAKE_COMMAND} -E echo "----------------------------------------------------------"
+    COMMAND ${CMAKE_COMMAND} -E echo "--- WARNING: Using dummy PSK! Only suitable for        ---"
+    COMMAND ${CMAKE_COMMAND} -E echo "--- development. Set NET_SAMPLE_PSK_HEADER_FILE to use ---"
+    COMMAND ${CMAKE_COMMAND} -E echo "--- own pre-shared key.                                ---"
+    COMMAND ${CMAKE_COMMAND} -E echo "----------------------------------------------------------"
```
How feasible would it be to add a custom command to generate the PSK every time?

As soon as user wanted to avoid their key to constantly change, they'd provide their own?

josuah reviewed (commented):
Some comments in hope this helps moving the review forward.

I am curious about why this got submitted to Zephyr rather than being a stand-alone module?

Prometheus is one particular vendor HTTP API (although, now maintained by several different companies).

Maybe it would be interesting to have it built as an OpenMetrics library?
https://en.wikipedia.org/wiki/Prometheus_(software)#Standardization_into_OpenMetrics

Thank you for all this work! :)

rlubos reviewed (commented):
Looks mostly good.

@mustafaabdullahk The merge window is now open again, so we can aim to finalize this. But it needs a rebase and remaining CI issues need to be addressed.

mustafaabdullahk commented on include/zephyr/net/prometheus/collector.h:
```diff
@@ -0,0 +1,77 @@
+/**
+ * @file
+ *
+ * @brief Prometheus collector APIs.
```
Yes i check the document(https://docs.zephyrproject.org/3.2.0/project/documentation.html) and added.

mustafaabdullahk commented on tests/net/lib/prometheus/collector/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+tests:
```
Added each test folder.

mustafaabdullahk commented on tests/net/lib/prometheus/counter/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+tests:
```
Added each test folder same in others :)

mustafaabdullahk commented:
> I wonder, shall we add a new entry for the library/sample in `MAINTAINERS.yml` file?

How should i add maintainer section truthly?

rlubos commented:
> > I wonder, shall we add a new entry for the library/sample in `MAINTAINERS.yml` file?
> 
> How should i add maintainer section truthly?

Have it been settled whether the library will end up in `subsys/mgmt` or `subsys/net`? In any case though, it's just a matter of creating a new entry in the `MAINTATINERS.yml` file, like for example this one:
https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml#L2655

jukkar commented on MAINTAINERS.yml:
```diff
@@ -2724,10 +2724,12 @@ Networking:
     - pdgendt
     - tbursztyka
     - ssharks
+    - mustafaabdullahk
```
I am not sure this is a good idea, now you will be start to get review requests for all the network PRs.
Probably better to create a separate Prometheus section and mark you as a maintainer there.

mustafaabdullahk replied:
Updated with your suggestions. Created Prometheus section under networking and marked as far as i understand.

jukkar commented on MAINTAINERS.yml:
```diff
@@ -2724,10 +2724,12 @@ Networking:
     - scripts/net/
     - drivers/net/
     - include/zephyr/net/
+    - include/zephyr/net/prometheus
```
The directory needs to have `/` at the end

jukkar commented on samples/net/prometheus/src/main.c:
```diff
@@ -0,0 +1,224 @@
+	static char print_str[MAX_TEMP_PRINT_LEN];
+	enum http_method method = client->method;
+	static size_t processed;
+	static bool replied = false;
```
No need to set static variables as false as they are that by default (CI complains about this)

jukkar commented on subsys/net/lib/prometheus/collector.c:
```diff
@@ -0,0 +1,145 @@
+		}
+	}
+
+	LOG_DBG("Counter not found %s", name);
```
Should this read `Gauge not found` ?

jukkar commented on subsys/net/lib/prometheus/collector.c:
```diff
@@ -0,0 +1,145 @@
+		}
+	}
+
+	LOG_DBG("Counter not found %s", name);
```
And here `Histogram not found`?

Suggestion for the these error debug prints
```
LOG_DBG("%s %s not found", "Histogram", name);
```
And just change the type of the item not found. This way the base string is shared between different functions and no new copy is stored in flash.

mustafaabdullahk replied:
Good point. Missed out. Will fix and push right now.

jukkar commented on subsys/net/lib/prometheus/collector.c:
```diff
@@ -0,0 +1,145 @@
+		}
+	}
+
+	LOG_DBG("Counter not found %s", name);
```
ditto

jukkar commented on subsys/net/lib/prometheus/formatter.c:
```diff
@@ -0,0 +1,226 @@
+			if (!write_metric_to_buffer(buffer + written, buffer_size - written,
+						    "# HELP %s %s\n", metric->name,
+						    metric->description)) {
+				LOG_WRN("Error writing to buffer");
```
Are these errors or warnings, the string indicates an error so should we use `LOG_ERR` here and in other warning prints below.

rlubos commented on MAINTAINERS.yml:
```diff
@@ -2926,6 +2926,28 @@ Networking:
+  labels:
+    - "area: Prometheus"
+  tests:
+    prometheus.collector
```
Wouldn't a single `- prometheus` entry enought to cover all test subcategories? Missing `-` also seems to be the case for the failing CI.

mustafaabdullahk replied:
Yes you'r right. Wrote to similiar entry but i removed and just keep 'prometheus'.

rlubos commented on include/zephyr/net/prometheus/collector.h:
```diff
@@ -0,0 +1,87 @@
+ * @defgroup prometheus_collector Prometheus Collector API
+ * @since 3.8.0
+ * @version 0.1.0
+ * @ingroup prometheus
```
I think we still need to `@defgroup prometheus` group (`@ingroup networking`) so that the APIs are visible. 

Also note, that with the following approach, each `@defgroup` will end up as a separate entry in the API summary page:
https://docs.zephyrproject.org/latest/develop/api/overview.html

I'll leave the decision up to you, but IMO we could simply have a single `@defgroup` entry:
```
 * @defgroup prometheus Prometheus API
 * @since 3.8.0
 * @version 0.1.0
 * @ingroup networking
```
And then just use the following in the remaining headers:
```
 * @addtogroup prometheus
```

mustafaabdullahk replied:
Missed out documents. Agree with you and updated your suggestions. Defined into collector.h file and just added '@addtogroup' field to other files.

rlubos reviewed (dismissed):
Just some remaining nits.

jukkar commented:
@mustafaabdullahk there is some compliance issues reported by CI, could you check those?

Please ignore the der binary file report, we will ask exception to this when the PR is accepted and ready to be merged.
```
Error: No binary files allowed.

 Binary file not allowed: samples/net/prometheus/src/ca.der
 Binary file not allowed: samples/net/prometheus/src/https-server-cert.der
 Binary file not allowed: samples/net/prometheus/src/https-server-key.der
 Binary file not allowed: samples/net/prometheus/src/server.der
 Binary file not allowed: samples/net/prometheus/src/server_privkey.der
Compliance error, check for error messages in the "Run Compliance Tests" step
You can run this step locally with the ./scripts/ci/check_compliance.py script.
Error: Process completed with exit code 1.
```

jukkar commented on include/zephyr/net/prometheus/collector.h:
```diff
@@ -0,0 +1,87 @@
+ * @brief Prometheus collector APIs.
+ *
+ * @defgroup prometheus Prometheus API
+ * @since 3.8.0
```
I checked the rendered documentation, and it seems this should be written as `@since 3.8` as now the doc output says `v3.8.0.0` which is not correct.
See https://builds.zephyrproject.io/zephyr/pr/72804/docs/develop/api/overview.html

mustafaabdullahk replied:
Fixed it. That issue related one file and update it as 3.8.

kartben commented on include/zephyr/net/prometheus/collector.h:
```diff
@@ -0,0 +1,87 @@
+ * @brief Prometheus collector APIs.
+ *
+ * @defgroup prometheus Prometheus API
+ * @since 3.8
```
There's no such thing as Zephyr 3.8 :)

```suggestion
 * @since 4.0
```

kartben commented on include/zephyr/net/prometheus/collector.h:
```diff
@@ -0,0 +1,87 @@
+ *
+ * @param _name The collector's name.
+ *
+ * @example
+ *
+ * PROMETHEUS_COLLECTOR_DEFINE(test_collector);
```
This is not the right use for this command 
https://www.doxygen.nl/manual/commands.html#cmdexample

In any case, this is not really needed as it's not adding any information IMO


```suggestion
```

kartben commented on include/zephyr/net/prometheus/counter.h:
```diff
@@ -0,0 +1,75 @@
+ * @param _name The channel's name.
+ * @param _detail The metric base.
+ *
+ * @example
```
Again not the right tag
You probably just want to use `@code`. Applies elsewhere in the PR

kartben replied:
See why it's wrong here: https://builds.zephyrproject.io/zephyr/pr/72804/docs/doxygen/html/structprometheus__counter.html

kartben commented on include/zephyr/net/prometheus/counter.h:
```diff
@@ -0,0 +1,75 @@
+ * @brief Increment the value of a Prometheus counter metric
+ * Increments the value of the specified counter metric by one.
+ * @param counter Pointer to the counter metric to increment.
+ *
+ * * Example *
+ * prometheus_counter_inc(counter);
```
```suggestion
```

kartben commented on include/zephyr/net/prometheus/label.h:
```diff
@@ -0,0 +1,42 @@
+ * @{
+ */
+
+/* maximum length of label key */
+#define MAX_LABEL_KEY_LENGTH   16
+/* maximum length of label value */
+#define MAX_LABEL_VALUE_LENGTH 16
+/* maximum namber of labels per metric */
+#define MAX_LABELS_PER_METRIC  5
+
```
Should this be more uniquely named? `MAX_PROMETHEUS_...`

kartben commented on include/zephyr/net/prometheus/metric.h:
```diff
@@ -0,0 +1,66 @@
+ * * See https://prometheus.io/docs/concepts/metric_types
+ */
+enum prometheus_metric_type {
+	/** Prometheuc Counter */
```
```suggestion
	/** Prometheus Counter */
```

kartben commented on include/zephyr/net/prometheus/metric.h:
```diff
@@ -0,0 +1,66 @@
+	struct prometheus_label labels[MAX_LABELS_PER_METRIC];
+	/** Number of labels associated with the Prometheus metric. */
+	int num_labels;
+	/** Add any other necessary fields */
```
```suggestion
	/* Add any other necessary fields */
```

kartben commented on samples/net/prometheus/Kconfig:
```diff
@@ -0,0 +1,41 @@
+mainmenu "Prometheus sample application"
+
+config NET_SAMPLE_HTTP_SERVICE
+	bool "Enable http service"
```
Nit: spell HTTP and HTTPS in uppercase

kartben commented on samples/net/prometheus/README.rst:
```diff
@@ -0,0 +1,87 @@
+Zephyr Prometheus Example
+=========================
```
Use `.. zephyr:code-sample::` directive and don't forget to flag `prometheus` as part of the "`relevant_apis`"

mustafaabdullahk replied:
Updated with following block:

```rst
.. zephyr-app-commands::
   :zephyr-app: samples/net/prometheus
   :board: <board to use>
   :conf: <config file to use>
   :goals: build
   :compact:
```

kartben commented on samples/net/prometheus/README.rst:
```diff
@@ -0,0 +1,87 @@
+
+To build and run the application:
+
+.. code-block:: bash
+
+   $ west build -p auto -b <board_to_use> -t run samples/net/prometheus
```
Use standard zephyr-app-commands directive (see other samples) instead

mustafaabdullahk replied:
Updated with following block:

```rst
.. zephyr-app-commands::
   :zephyr-app: samples/net/prometheus
   :board: <board to use>
   :conf: <config file to use>
   :goals: build
   :compact:
```

kartben commented on samples/net/prometheus/README.rst:
```diff
@@ -0,0 +1,87 @@
+
+This sample application demonstrates the use of the ``prometheus`` library.
+This library provides prometheus client library(pull method) implementation.
+By integrating this  library into your code, you can expose internal metrics
```
```suggestion

By integrating this library into your code, you can expose internal metrics
```

kartben commented:
> Please ignore the der binary file report, we will ask exception to this when the PR is accepted and ready to be merged.

It would be really great if we could avoid adding binary files to the tree. Can't these be generated by the user?

kartben reviewed (dismissed):
Some initial documentation related comments but I might need another pass and will try to actually test the PR too.

jukkar commented:
> It would be really great if we could avoid adding binary files to the tree. Can't these be generated by the user?

That is certainly possible but requires changes to the current workflow in the networking samples that need TLS support. For the user it is quite convenient if the certificates are already there when using the net sample.
One option could be to store only one set of certificates that all the networking samples could use, this way updating the certs could be done in one place.

rlubos commented:
> > It would be really great if we could avoid adding binary files to the tree. Can't these be generated by the user?
> 
> That is certainly possible but requires changes to the current workflow in the networking samples that need TLS support. For the user it is quite convenient if the certificates are already there when using the net sample. One option could be to store only one set of certificates that all the networking samples could use, this way updating the certs could be done in one place.

I don't think users can generate own certificates in all cases, various samples acting as TLS client often use official CA certificates to connect to public servers, which need to be somehow provided to the sample. And DER format (although considered binary) seems to be the most memory-efficient way to provide certs (it's not that we'll be "patching" those files anyway, typically they're replaced as a whole).

But for the TLS server cases, we could have common certificates for all samples (possibly under `samples/net/common/`). Not sure such a cleanup applies to this PR though.

jukkar commented:
@carlescufi could you force push this, the CI complains about binary certificate files so this needs higher powers.

kartben commented:
> @kartben you have -1 to this, could you recheck?

I think https://github.com/zephyrproject-rtos/zephyr/pull/72804#discussion_r1715839665 still applies but maybe it's not enough to justify a -1

btw, @mustafaabdullahk  in the future please don't mark comments as resolved until the original commenter had a chance to check for themselves.
I will trust that you indeed resolved all the comments for the sake of unblocking the PR, but I'd have much preferred being able to easily be able to check myself :) Please see https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers

mustafaabdullahk commented:
> btw, @mustafaabdullahk in the future please don't mark comments as resolved until the original commenter had a chance to check for themselves.
> I will trust that you indeed resolved all the comments for the sake of unblocking the PR, but I'd have much preferred being able to easily be able to check myself :) Please see

I was really wondering 'mark as resolved' button who should mark. I will do your suggestion in future PRs :). Thank you

jukkar commented:
> I was really wondering 'mark as resolved' button who should mark.

Indeed, even as I am a long time contributor, I have been wondering that myself too. I usually mark things resolved in order to keep track what I have fixed in my own PRs, but I suppose it can be useful info also for the reporter. Too bad there isn't two buttons, one for reporter and one for the developer, to track these things.

kartben commented:
@mustafaabdullahk To be clear you might very well have done the right thing as per the document I linked, if you applied all my suggestions verbatim, so no worries. It's just that my recollection is that I made some suggestions that were not directly "suggested changes" that could just be applied "as-is" so was definitely expecting to have to come back to the associated comments to check how you applied the suggestions :) Anyhow, again, no worries, and thanks for the great PR!

kartben reviewed (commented):
actually, @mustafaabdullahk, here's one good example of why leaving the comments open is useful :)

mustafaabdullahk commented on samples/net/prometheus/README.rst:
```diff
@@ -0,0 +1,93 @@
+   :name: Prometheus Sample
+   :relevant-api: http_service http_server tls_credentials prometheus
+
+   Implement an Prometheus Metric Server demonstrating various metric types.
```
Yes, we'r start with good example :). Solved it.

nashif commented on MAINTAINERS.yml:
```diff
@@ -2987,6 +2987,23 @@ Networking:
+"Networking: Prometheus":
+  status: maintained
+  maintainers:
+    - mustafaabdullahk
```
this is not going to work, because you are not part of the project and did not go through the process described here: https://docs.zephyrproject.org/latest/project/project_roles.html

So I suggest to remove yourself as maintainer, maybe drop this entry completely for now and add it later with yourself as maintainer after gone through the above process.

mustafaabdullahk replied:
Removed to your suggestion. I will write maybe after gone through the above process.

bjarki-andreasen replied:
You should just drop the last commit :)

mustafaabdullahk replied:
A tiny misunderstanding @nashif  :)

josuah commented:
The CI is complaining about the presence of binary files.
> [Error: No binary files allowed](https://github.com/zephyrproject-rtos/zephyr/actions/runs/10959894887/job/30433240586?pr=72804#step:11:45).
> Binary file not allowed: samples/net/prometheus/src/ca.der
> Binary file not allowed: samples/net/prometheus/src/https-server-cert.der
> Binary file not allowed: samples/net/prometheus/src/https-server-key.der
> Binary file not allowed: samples/net/prometheus/src/server.der
> Binary file not allowed: samples/net/prometheus/src/server_privkey.der

Propositions:
- include a PEM file and convert it into DER at build-time,
- store DER as base64/hexdump
- generate new certificates at build-time.
- allow `.der` files

commit 7b57d3c5bbd151dfa4d94f3434488b4595f07022
Author: @fabiobaltieri
Date:   Mon Nov 21 13:43:55 2022 +0000

    scripts: compliance: check for binary files

    Add a compliance check that fails if the diff is adding any binary file
    to prevent committing them by mistake.

    Only check for added files and explicitly allow image files in doc/ and
    boards/.

    Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

jukkar commented:
> The CI is complaining about the presence of binary files.
> 
> > [Error: No binary files allowed](https://github.com/zephyrproject-rtos/zephyr/actions/runs/10959894887/job/30433240586?pr=72804#step:11:45).
> > Binary file not allowed: samples/net/prometheus/src/ca.der
> > Binary file not allowed: samples/net/prometheus/src/https-server-cert.der
> > Binary file not allowed: samples/net/prometheus/src/https-server-key.der
> > Binary file not allowed: samples/net/prometheus/src/server.der
> > Binary file not allowed: samples/net/prometheus/src/server_privkey.der
> 
> Propositions:
> 
>     * include a PEM file and convert it into DER at build-time,
> 
>     * store DER as base64/hexdump
> 
>     * generate new certificates at build-time.
> 
>     * allow `.der` files
> 
>
It has been agreed that for the binary certificate files, we ask exception to merge. So no worries for this.

jukkar commented:
@nashif / @carlescufi please force push to override the binary file check issue

josuah commented on samples/net/prometheus/src/main.c:
```diff
@@ -0,0 +1,222 @@
+
+	prometheus_format_exposition(prom_context.collector, recv_buffer, sizeof(recv_buffer));
+
+	/**
```
Comments starting with `/**` are going to be picked-up by Doxygen.
If this was meant as a an explanation of implementation, you can use `/*` instead.

josuah commented on include/zephyr/net/prometheus/counter.h:
```diff
@@ -0,0 +1,74 @@
+	/** Base of the Prometheus counter metric */
+	struct prometheus_metric *base;
+	/** Value of the Prometheus counter metric */
+	unsigned long value;
```
For an event happening every 10 millisecond, AFAIU, where `unsigned long` is 32-bit, this would overflow every 1.3 year? Someone migrating from a 64-bit (overflow every  to a 32-bit platform could get surprised.


`

josuah replied:
Maybe this is not a matter to solve by this library but by the user?

If wishing to address this, using `uint64_t` or `uint32_t` (and documenting it) could help eventually.
With `uint32_t`, checking for overflow and reporting it could be good addition.

Not sure `uint64_t` will ever overflow if incrementing by just 1 every time:
```python
>>> 0xffffffffffffffff /  (1000000000 * 3600 * 24 * 365)
overflow every 584.942417355072 years if logging every nanosecond
```

josuah replied:
https://en.cppreference.com/w/cpp/language/types

josuah commented on subsys/net/lib/prometheus/formatter.c:
```diff
@@ -0,0 +1,226 @@
+static int write_to_buffer(char *buffer, size_t buffer_size, const char *str)
+{
+	/* helper function to write formatted string to buffer, handling buffer size */
+	int len = snprintf(buffer + strlen(buffer), buffer_size - strlen(buffer), "%s", str);
```
There is an opportunity to use wrapper functions such as `net_buf` instead of rolling out buffer handling, which is a library already used by other network interfaces, so the code can be reused.

In absence of `strlcat()`, `snprintf()` sounds like the way to go.

josuah commented on subsys/net/lib/prometheus/formatter.c:
```diff
@@ -0,0 +1,226 @@
+	/* helper function to write formatted string to buffer, handling buffer size */
+	int len = snprintf(buffer + strlen(buffer), buffer_size - strlen(buffer), "%s", str);
+
+	if (len >= 0 && len < buffer_size - strlen(buffer)) {
+		return 1;
+	}
```
Zephyr functions typically return a negative error code with a value from `errno.h`.
For instance, `-ENOMEM` or `-ENAMETOOLONG`.

josuah commented on subsys/net/lib/prometheus/formatter.c:
```diff
@@ -0,0 +1,226 @@
+		return 1;
+	}
+
+	/* error (insufficient buffer size) */
+	return 0;
```
Zephyr functions typically `return 0` on success and a negative value on failure, like libc syscalls.

A typical idiom is:
```
int ret;

ret = function_call_that_might_fail();
if (ret < 0) {
        LOG_ERR("optional error message");
        return ret;
}

return 0;
```

`int err;` and `if (err) { return err; };` is another variant used.

josuah replied:
You would need to propagate this change all the way down:
```diff
-if (!function(args)) {
-        return 0;
-}
+int err;
+
+err = function(args);
+if (err < 0) {
+        return 0;
+}
```

It is a bit more verbose, but a very frequent idiom.

josuah replied:
Thank you for the effort of changing it everywhere!

Some last minor edit at this line:
```suggestion
	return -ENOBUFS;
```

`NOBUFS` is for "No buffer space available". Maybe `-ENOMEM` or something else is also suited.

josuah commented on samples/net/prometheus/src/main.c:
```diff
@@ -0,0 +1,222 @@
+
+	(void)memset(recv_buffer, 0, sizeof(recv_buffer));
+
+	prometheus_format_exposition(prom_context.collector, recv_buffer, sizeof(recv_buffer));
```
This function can fail but it returns `void`, and is called in a function that returns `int`.

How about taking advantage of your existing error checking effort to connect the error reporting further? :)

```suggestion
	err = prometheus_format_exposition(prom_context.collector, recv_buffer, sizeof(recv_buffer));
	if (err < 0) {
		return err;
	}
```

josuah commented on samples/net/prometheus/src/main.c:
```diff
@@ -0,0 +1,222 @@
+		return rc;
+	}
+
+	memcpy(recv_buffer + strlen(recv_buffer), "\r\n", 2);
```
What happens if `strlen(recv_buffer) == sizeof(recv_buffer)` here?
One more point for [`net_buf`](https://docs.zephyrproject.org/latest/services/net_buf/index.html)

josuah replied:
That being said, it is also possible that `recv_buffer` is large enough for anything coming from here: https://github.com/zephyrproject-rtos/zephyr/pull/72804/files/7c6366cd4d818e68a6f18cf31a108ac979dc74c9#diff-39b74f8d5670ba8f827d9fbc421275e870d519da3b2121771edb00d6e7157c27R85

josuah commented on subsys/net/lib/prometheus/formatter.c:
```diff
@@ -0,0 +1,226 @@
+			const struct prometheus_gauge *gauge =
+				(const struct prometheus_gauge *)prometheus_collector_get_metric(
+					collector, metric->name);
+			;
```
Maybe this is from a previous refactoring?

josuah commented on subsys/net/lib/prometheus/gauge.c:
```diff
@@ -0,0 +1,27 @@
+{
+	if (value < 0) {
+		LOG_ERR("Invalid value");
+		return;
```
How about making the logger return an error code?
```suggestion
		return -EINVAL;
```

josuah commented on samples/net/prometheus/src/main.c:
```diff
@@ -0,0 +1,222 @@
+	}
+
+	memcpy(recv_buffer + strlen(recv_buffer), "\r\n", 2);
+	memcpy(buffer, recv_buffer, strlen(recv_buffer));
```
What happens if buffer's `len` is 128 and `strlen(recv_buffer)` is 234?
One more point for [`net_buf`](https://docs.zephyrproject.org/latest/services/net_buf/index.html) :)

josuah reviewed (changes requested):
> EDIT: to avoid blocking this PR, the discussion about using `net_buf` for string handling in network protocols is going here:
> - https://github.com/zephyrproject-rtos/zephyr/issues/78799

Thank you for this extra round of review! The API is comfortable: not a lot of effort to define a new metric type.

~~I wonder what is the best way to do string handling in Zephyr... I suggested `net_buf` because it is already there on network libraries, but maybe there need to be some `net_buf_add_str()`...~~ [see #78799]

~~Maybe in the meantime~~ it is possible to ~~use `net_buf_add_mem()` or~~ keep using the `string.h` functions...

~~I do not see `net_buf` used anywhere in Zephyr for string handling yet. Maybe I am missing something?~~

mmahadevan108 commented:
I am curious to know the answer for this comment .https://github.com/zephyrproject-rtos/zephyr/pull/72804#pullrequestreview-2138939367

Can this be a `module` inside the `lib` folder?

jukkar commented:
There is one compliance error that we could fix
```
 OPEN_BRACE: that open brace { should be on the previous line
File:samples/net/prometheus/src/main.c
Line:76
Compliance error, check for error messages in the "Run Compliance Tests" step
You can run this step locally with the ./scripts/ci/check_compliance.py script.
Error: Process completed with exit code 1.
```

mustafaabdullahk commented:
> There is one compliance error that we could fix
> 
> ```
>  OPEN_BRACE: that open brace { should be on the previous line
> File:samples/net/prometheus/src/main.c
> Line:76
> Compliance error, check for error messages in the "Run Compliance Tests" step
> You can run this step locally with the ./scripts/ci/check_compliance.py script.
> Error: Process completed with exit code 1.
> ```

I just apply clang-format. Check complience pipeline give error following type.
```c
struct http_resource_detail_dynamic dyn_resource_detail = {
	.common =
		{
			.type = HTTP_RESOURCE_TYPE_DYNAMIC,
			.bitmask_of_supported_http_methods = BIT(HTTP_GET),
			.content_type = "text/plain",
		},
	.cb = dyn_handler,
	.user_data = NULL,
};
```
but if i move to bracket previous line like following i give warning (https://github.com/zephyrproject-rtos/zephyr/actions/runs/11113829987). Which one is true?:

```c
struct http_resource_detail_dynamic dyn_resource_detail = {
	.common = {
			.type = HTTP_RESOURCE_TYPE_DYNAMIC,
			.bitmask_of_supported_http_methods = BIT(HTTP_GET),
			.content_type = "text/plain",
		},
	.cb = dyn_handler,
	.user_data = NULL,
};
```

jukkar commented:
As Pieter said, just ignore the clang- format suggestions for now.

mustafaabdullahk commented:
@josuah gentle ping :)

josuah commented on subsys/net/lib/prometheus/histogram.c:
```diff
@@ -0,0 +1,39 @@
+		if (value <= histogram->buckets[i].upper_bound) {
+			/* increment count for the bucket */
+			histogram->buckets[i].count++;
+			break;
```
This means that values out of range will silently be ignored instead of triggering an error or land in the top bucket.

If you wished to avoid it, you could change `break;` by `return 0;` and at the bottom of the function set maybe a `return -ERANGE;` or `-EOVERFLOW` or `-ENOTSUP` or something more fitting.

No need to change that if that was the expected behavior though!
In which case, it is also possible to add a `LOG_DBG()` to help someone using the lib to understand "why is my value never added?".

mustafaabdullahk replied:
It is `Apdex score` related issue but you are right, user want to know and debug values in some cases. Added debug message for condition branch.

josuah replied:
Insightful, thank you for the reference and modification!

josuah commented on tests/net/lib/prometheus/counter/src/main.c:
```diff
@@ -0,0 +1,46 @@
+	zassert_true(test_counter_m.value == 0, "Counter value is not 0");
+
+	ret = prometheus_counter_inc(&test_counter_m);
+	zassert_true(ret == 0, "Error incrementing counter");
```
I recently learned about [`zassert_ok(ret, "msg");`](https://docs.zephyrproject.org/latest/doxygen/html/group__ztest__assert.html#gade40e2ec78ec813739e7725524fae7f0).
Same observation applies to other tests.

mustafaabdullahk replied:
Me too i recently learned and applied to your suggestions each test cases :)

josuah commented on tests/net/lib/prometheus/counter/src/main.c:
```diff
@@ -0,0 +1,46 @@
+{
+	int ret;
+
+	zassert_true(test_counter_m.value == 0, "Counter value is not 0");
```
You can also use `zassert_equal(test_counter_m.value, 0, "...")`.
Same observation applies to other tests.

josuah reviewed (changes requested):
Thank you for your patience, I was traveling but I am back now.
Some very minor changes proposed, but it looks very close now.
Thank you for all the iterations!

josuah commented on subsys/net/lib/prometheus/Kconfig:
```diff
@@ -0,0 +1,31 @@
+	help
+	  Maximum number of metrics that can be registered.
+
+config PROMETHEUS_METRIC_FORMAT_BUFFER_SIZE
+	int "Exposition buffer size"
+	default 256
+	help
+	  Size of the buffer that holds the exposition data.
```
See above if interested in a small optimization which permits to remove this temporary buffer.

mustafaabdullahk replied:
Removed with updated related `write_to_buffer` parts.

josuah commented on subsys/net/lib/prometheus/formatter.c:
```diff
@@ -0,0 +1,242 @@
+
+static int write_metric_to_buffer(char *buffer, size_t buffer_size, const char *format, ...)
+{
+	/* helper function to write formatted metric to buffer */
+	char temp_buffer[CONFIG_PROMETHEUS_METRIC_FORMAT_BUFFER_SIZE];
+	va_list args;
+
+	va_start(args, format);
+	vsnprintf(temp_buffer, sizeof(temp_buffer), format, args);
+	va_end(args);
+
+	return write_to_buffer(buffer, buffer_size, temp_buffer);
```
I think it is possible to combine the `vsnprintf()` from here and the `snprintf()` from `write_to_buffer()` which is used only once in the end...

```suggestion
	va_list args;
	size_t len = strlen(buffer);

	if (len >= buffer_size) {
		return -ENOMEM;
	}
	buffer += len;
	buffer_size -= len;

	va_start(args, format);
	len = vsnprintf(buffer, buffer_size, format, args);
	va_end(args);
	if (len >= buffer_size) {
		return -ENOMEM;
	}

	return 0;
```

This would allow to remove `CONFIG_PROMETHEUS_METRIC_FORMAT_BUFFER_SIZE` from Kconfig.
This is `>=` instead of `>` because `vsnprintf()` does not count the terminating `'\0'`.
This also always return `-error_number` on error and `0` on success.
That is just a small optimization though...

mustafaabdullahk replied:
It is good suggestion for that case. I did applied :)

mustafaabdullahk committed:
net: add initial prometheus client library

The library provides Prometheus metrics
types, collector and exposion formatter.
The library isn't thread-safe for now.
The next first pull request will support
that. Can be use exposion formatted
output with Zephyr Http server.

Signed-off-by: Mustafa Abdullah Kus <mustafa.kus@sparsetechnology.com>

mustafaabdullahk committed:
samples: net: add prometheus library sample application

The sample uses the Zephyr HTTP server library
and demonstrates the Prometheus metric
server node. Prometheus client library
runs as a pull method. The sample contains
the HTTP request counter and increases
when refresh path of '/metrics' from the browser.

Signed-off-by: Mustafa Abdullah Kus <mustafa.kus@sparsetechnology.com>

mustafaabdullahk committed:
test: net: lib: add prometheus library tests

Tests for prometheus library support.
It contains integration and unit test
each metric types.

Signed-off-by: Mustafa Abdullah Kus <mustafa.kus@sparsetechnology.com>

josuah commented on subsys/net/lib/prometheus/formatter.c:
```diff
@@ -0,0 +1,237 @@
+				 size_t buffer_size)
+{
+	int ret;
+	int written = 0;
```
Maybe it is thanks to a previous refactoring... It looks like `written` is not updated and is always 0, so could be removed.

josuah reviewed (approved):
As far as I can see, everything got fixed and addressed!
One minor suggestion, but does not hurt in any way.
Thank you for all the changes!

bjarki-andreasen commented:
congratulations! :champagne:
